python/src/opendp/polars.py

```diff
@@ -190,3 +196,180 @@ def lazy(self):
+
+        def __getattribute__(self, name):
+            # Re-route all possible attribute access to self._lf_plan.
+            # __getattribute__ is necessary because __getattr__ cannot intercept calls to inherited methods
+
+            lf_plan = object.__getattribute__(self, "_lf_plan")
+            query = object.__getattribute__(self, "_query")
+            attr = getattr(lf_plan, name, None)
+
+            # If not a valid attribute on self._lf_plan, then don't re-route
+            if attr is None:
+                return object.__getattribute__(self, name)
+
+            # any callable attributes (like .with_columns or .select) will now also wrap their outputs in a LazyFrameQuery
+            if callable(attr):
+
+                def wrap(*args, **kwargs):
+                    out = attr(*args, **kwargs)
+
+                    # We keep the query plan void of data anyways,
+                    # so running the computation doesn't affect privacy.
+                    # This isn't meant to cover all possible APIs that may execute the query,
+                    #    but it does give a simple sanity check in the obvious cases.
+                    if isinstance(out, DataFrame):
+                        raise ValueError(
+                            "You must call `.release()` before executing a query."
+                        )
+
+                    # re-wrap any lazy outputs to keep the conveniences afforded by this class
+                    if isinstance(out, (LazyGroupBy, LazyFrame)):
+                        return LazyFrameQuery(out, query)
+
+                    return out
+
+                return wrap
+            return attr
+
+        def resolve(self) -> Measurement:
+            """Resolve the query into a measurement."""
+            from opendp.context import PartialChain
+
+            # access attributes of self without getting intercepted by Self.__getattribute__
+            lf_plan = object.__getattribute__(self, "_lf_plan")
+            query = object.__getattribute__(self, "_query")
+            input_domain, input_metric = query._chain
+
+            # create an updated query and then resolve it into a measurement
+            return query.new_with(
+                chain=PartialChain(
+                    lambda s: make_private_lazyframe(
+                        input_domain=input_domain,
+                        input_metric=input_metric,
+                        output_measure=query._output_measure,
+                        lazyframe=lf_plan,
+                        global_scale=s,
+                    )
+                )
+            ).resolve()
+
+        def release(self) -> OnceFrame:
+            """Release the query. The query must be part of a context."""
+            query = object.__getattribute__(self, "_query")
+            resolve = object.__getattribute__(self, "resolve")
+            return query._context(resolve())  # type: ignore[misc]
+
+except ImportError:
```

Comment:
I worry a little about the really long try block: We think we know where the `ImportError` would come from... but if one popped up later on in the blcok, and we ended up here, it would be confusing.

Reply:
I've trimmed this block down as much as possible: Sphinx check, import, class definition. Anything raised inside the class definition won't get caught by the exception. I can't seem to finagle it so that the class definition is outside but Pylance still understands it.